src/crimson/os/seastore/record_scanner.cc

```diff
@@ -0,0 +1,239 @@
+		    seastar::stop_iteration>(seastar::stop_iteration::no);
+		});
+	      });
+	  });
+	} else {
+	  assert(!cursor.pending_record_groups.empty());
+	  auto &next = cursor.pending_record_groups.front();
+	  return read_validate_data(next.offset, next.header
+	  ).safe_then([this, FNAME, &budget_used, &cursor, &handler, &next](auto valid) {
+	    if (!valid) {
+	      INFO("complete at {}, invalid record group data at {}, {}",
+		   cursor, next.offset, next.header);
+	      cursor.pending_record_groups.clear();
+	      return scan_valid_records_ertr::now();
+	    }
+            return consume_next_records(cursor, handler, budget_used);
+	  });
+	}
+      }().safe_then([=, &budget_used, &cursor] {
+	if (cursor.is_complete() || budget_used >= budget) {
+	  DEBUG("finish at {}, budget_used={}, budget={}",
+                cursor, budget_used, budget);
+	  return seastar::stop_iteration::yes;
+	} else {
+	  return seastar::stop_iteration::no;
+	}
+      });
+    }).safe_then([retref=std::move(retref)]() mutable -> scan_valid_records_ret {
+      return scan_valid_records_ret(
+	scan_valid_records_ertr::ready_future_marker{},
+	std::move(*retref));
+    });
+}
+
+RecordScanner::read_validate_record_metadata_ret
+RecordScanner::read_validate_record_metadata(
+  scan_valid_records_cursor &cursor,
+  segment_nonce_t nonce)
+{
+  LOG_PREFIX(RecordScanner::read_validate_record_metadata);
+  paddr_t start = cursor.seq.offset;
+  auto block_size = cursor.get_block_size();
+  if (get_segment_off(cursor.seq.offset) + block_size > get_segment_end_offset(cursor.seq.offset)) {
+    DEBUG("failed -- record group header block {}~4096 > segment_size {}",
+      start, get_segment_end_offset(cursor.seq.offset));
+    return read_validate_record_metadata_ret(
+      read_validate_record_metadata_ertr::ready_future_marker{},
+      std::nullopt);
+  }
+  TRACE("reading record group header block {}~4096", start);
+  return read(start, block_size
+  ).safe_then([=](bufferptr bptr) mutable
+              -> read_validate_record_metadata_ret {
+    bufferlist bl;
+    bl.append(bptr);
+    auto maybe_header = try_decode_records_header(bl, nonce);
+    if (!maybe_header.has_value()) {
+      return read_validate_record_metadata_ret(
+        read_validate_record_metadata_ertr::ready_future_marker{},
+        std::nullopt);
+    }
+
+    auto& header = *maybe_header;
+    if (header.mdlength < block_size ||
+        header.mdlength % block_size != 0 ||
+        header.dlength % block_size != 0 ||
+	(header.committed_to != JOURNAL_SEQ_NULL &&
+	get_segment_off(header.committed_to.offset) %
+	cursor.get_block_size() != 0) ||
+	(get_segment_off(cursor.seq.offset) + header.mdlength + header.dlength >
+	get_segment_end_offset(cursor.seq.offset))) {
+      ERROR("failed, invalid record group header {}", header);
+      return crimson::ct_error::input_output_error::make();
+    }
+
+    if (is_record_segment_seq_valid(cursor, header)) {
```

Comment:
According to the name, false should imply the record md is invalid and return nullopt.